### PR TITLE
PIM-5782: add sort order support on attribute selector

### PR DESCRIPTION
**Pull request by juliensnz**
**Description (for Contributor and Core Developer)**

Add sort order to the product export builder

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Y
| Added Behats                      | Y
| Changelog updated                 | N
| Review and 2 GTM                  | N
| Micro Demo to the PO (Story only) | N
| Migration script                  | NA
| Tech Doc                          | NA

### CODE REVIEW

File: src/Pim/Bundle/EnrichBundle/Normalizer/ChannelNormalizer.php [hunk 1]
```diff
@@ -66,6 +66,6 @@ protected function normalizeLocales($locales)
     {
         return array_map(function ($locale) {
             return $this->localeNormalizer->normalize($locale, 'json');
-        }, $this->collectionFilter->filterCollection($locales, 'pim.internal_api.locale.view'));
+        }, $this->collectionFilter->filterCollection($locales, 'pim.internal_api.locale.view')->toArray());
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Normalizer/ChannelNormalizer.php]: again?

File: src/Pim/Component/Connector/Writer/File/ProductColumnSorter.php [hunk 2]
```diff
@@ -35,9 +35,14 @@ public function __construct(
     /**
      * {@inheritdoc}
      */
-    public function sort(array $columns)
+    public function sort(array $columns, array $context = [])
     {
         $identifier = $this->productRepository->getIdentifierProperties()[0];
+
+        if (isset($context['filters']['structure']['attributes']) && !empty($context['filters']['structure']['attributes'])) {
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/File/ProductColumnSorter.php]: disagree
here you don't care about the export builder

your context should be specific to this class
just pass what you need

File: src/Pim/Bundle/ConnectorBundle/Resources/config/writers.yml [hunk 3]
```diff
@@ -384,4 +384,4 @@ services:
         arguments:
             - '@pim_connector.array_converter.flat_to_standard.product.field_splitter'
             - '@pim_catalog.repository.product'
-            - ['label']
```
Comment (jjanvier) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/writers.yml]: why did you remove label?
not used by the quick export? sure??

File: src/Pim/Component/Connector/Writer/File/ProductColumnSorter.php [hunk 4]
```diff
@@ -35,9 +35,14 @@ public function __construct(
...
+            return array_merge([$identifier], $this->firstDefaultColumns, $context['filters']['structure']['attributes']);
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/File/ProductColumnSorter.php]: I don't understand why you don't use the parent here
it seems you omit things

File: src/Pim/Bundle/EnrichBundle/Normalizer/ChannelNormalizer.php (hunk 1 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Normalizer/ChannelNormalizer.php]: No, it was a quick fix on this branch to be able to run the CI :) I made this commit yesterday before knowing that this solution was failing on EE. I will revert it, good catch.

File: src/Pim/Component/Connector/Writer/File/ProductColumnSorter.php (hunk 2 above)
Comment (juliensnz) [inline src/Pim/Component/Connector/Writer/File/ProductColumnSorter.php]: Not really possible: here we are in a PRODUCTColumnSorter but we pass the context in the FlatItemBufferFlusher. So I can not put PEB specific stuff in a generic class.

File: src/Pim/Bundle/ConnectorBundle/Resources/config/writers.yml (hunk 3 above)
Comment (juliensnz) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/writers.yml]: Yeah, you are right. I will have to duplicate all this shit :(

File: src/Pim/Component/Connector/Writer/File/ProductColumnSorter.php (hunk 4 above)
Comment (juliensnz) [inline src/Pim/Component/Connector/Writer/File/ProductColumnSorter.php]: the parent is just sorting the collection of code provided. But the goal of this PR is to not sort them so I don't see why we should call it

File: src/Pim/Component/Connector/Writer/File/ProductColumnSorter.php (hunk 2 above)
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/File/ProductColumnSorter.php]: you added this in the FlatItemBufferFlusher
so in the FlatItemBufferFlusher you have to keep only what's relevant as context for the Sorter

File: features/export/product-export-builder/export_products_with_attributes.feature [hunk 5]
```diff
@@ -62,6 +62,34 @@ Feature: Export products with only selected attributes
...
+  Scenario: Export products by selecting multiple attribute using the UI in a specific order
+    Given the following job "csv_footwear_product_export" configuration:
+      | filePath | %tmp%/product_export/product_export.csv |
+    When I am on the "csv_footwear_product_export" export job edit page
+    And I visit the "Content" tab
+    And I filter by "completeness" with operator "No condition on completeness" and value ""
+    And I select the following attributes to export lace_color and weather_conditions
+    And I press the "Save" button
+    Then I should not see the text "There are unsaved changes"
+    And I launch the export job
```
Comment (damien-carcel) [inline features/export/product-export-builder/export_products_with_attributes.feature]: When

File: features/export/product-export-builder/export_products_with_attributes.feature [hunk 6]
```diff
@@ -62,6 +62,34 @@ Feature: Export products with only selected attributes
...
+    And I wait for the "csv_footwear_product_export" job to finish
+    Then exported file of "csv_footwear_product_export" should contains the following headers:
+    """
+    sku;categories;enabled;family;groups;lace_color;weather_conditions
+    """
+    Then I am on the "csv_footwear_product_export" export job edit page
```
Comment (damien-carcel) [inline features/export/product-export-builder/export_products_with_attributes.feature]: When

File: features/export/product-export-builder/export_products_with_attributes.feature [hunk 7]
```diff
@@ -62,6 +62,34 @@ Feature: Export products with only selected attributes
...
+    And I visit the "Content" tab
+    And I filter by "completeness" with operator "No condition on completeness" and value ""
+    And I select the following attributes to export weather_conditions and lace_color
+    And I press the "Save" button
+    Then I should not see the text "There are unsaved changes"
+    And I launch the export job
+    And I wait for the "csv_footwear_product_export" job to finish
+    Then exported file of "csv_footwear_product_export" should contains the following headers:
```
Comment (damien-carcel) [inline features/export/product-export-builder/export_products_with_attributes.feature]: And

File: src/Pim/Component/Connector/Writer/File/ProductColumnSorter.php (hunk 2 above)
Comment (juliensnz) [inline src/Pim/Component/Connector/Writer/File/ProductColumnSorter.php]: seen IRL

**Comment by damien-carcel** [comment]
GTM once comments fixed.

File: features/export/product-export-builder/export_products_with_attributes.feature (hunk 7 above)
Comment (juliensnz) [inline features/export/product-export-builder/export_products_with_attributes.feature]: don't agree

**Comment by solivier** [comment]
GTM
